sock_trigger.c is missing

A build on master results in:

```
make  all-am
make[1]: Entering directory `/home/jsquyres/git/libfabric'
make[1]: *** No rule to make target `prov/sockets/src/sock_trigger.c', needed by `prov/sockets/src/src_libfabric_la-sock_trigger.lo'.  Stop.
```

--- a/Makefile.am
+++ b/Makefile.am
@@ -80,6 +80,7 @@ _sockets_files = \
 	prov/sockets/src/sock_msg.c \
 	prov/sockets/src/sock_rma.c \
 	prov/sockets/src/sock_atomic.c \
+	prov/sockets/src/sock_trigger.c \
 	prov/sockets/src/sock_util.h \
 	prov/sockets/src/indexer.c
 

--- a/man/fi_sockets.7.md
+++ b/man/fi_sockets.7.md
@@ -44,15 +44,12 @@ Key features include:
 
 # LIMITATIONS
 
-Sockets provider currently do not support triggered operations. This
-will be added in future releases.
-
-Sockets provider attempts to emulate the entire API set (except the
-triggered operations), including all defined options. In order to
-support development on a wide range of systems, it is implemented over
-TCP sockets. As a result, the performance numbers are lower compared
-to other providers implemented over high-speed fabric, and lower than
-what an application might see implementing to sockets directly.
+Sockets provider attempts to emulate the entire API set, including all
+defined options. In order to support development on a wide range of
+systems, it is implemented over TCP sockets. As a result, the
+performance numbers are lower compared to other providers implemented
+over high-speed fabric, and lower than what an application might see
+implementing to sockets directly.
 
 # SEE ALSO
 

--- a/prov/sockets/src/sock.h
+++ b/prov/sockets/src/sock.h
@@ -102,14 +102,14 @@
 			 FI_SOURCE | FI_READ | FI_WRITE | FI_RECV | FI_SEND | \
 			 FI_REMOTE_READ | FI_REMOTE_WRITE | \
 			 FI_RMA_EVENT | \
-			 FI_MORE | FI_FENCE)
+			 FI_MORE | FI_FENCE | FI_TRIGGER)
 
 #define SOCK_EP_MSG_CAP SOCK_EP_RDM_CAP
 
 #define SOCK_EP_DGRAM_CAP (FI_MSG | FI_TAGGED | \
 			   FI_NAMED_RX_CTX | FI_DIRECTED_RECV | \
 			   FI_MULTI_RECV | FI_SOURCE | FI_RECV | FI_SEND | \
-			   FI_MORE | FI_FENCE)
+			   FI_MORE | FI_FENCE | FI_TRIGGER)
 
 #define SOCK_EP_MSG_ORDER (FI_ORDER_RAR | FI_ORDER_RAW | FI_ORDER_RAS|	\
 			   FI_ORDER_WAR | FI_ORDER_WAW | FI_ORDER_WAS |	\
@@ -192,6 +192,43 @@ struct sock_domain {
 	struct fi_domain_attr attr;
 };
 
+struct sock_trigger {
+	uint8_t op_type;
+	size_t threshold;
+	struct dlist_entry entry;
+
+	struct fid_ep	*ep;
+	uint64_t flags;
+	
+	union {
+		struct {
+			struct fi_msg msg;
+			struct iovec msg_iov[SOCK_EP_MAX_IOV_LIMIT];
+		} msg;
+
+		struct {
+			struct fi_msg_tagged msg;
+			struct iovec msg_iov[SOCK_EP_MAX_IOV_LIMIT];
+		} tmsg;
+
+		struct {
+			struct fi_msg_rma msg;
+			struct iovec msg_iov[SOCK_EP_MAX_IOV_LIMIT];
+			struct fi_rma_iov rma_iov[SOCK_EP_MAX_IOV_LIMIT];
+		} rma;
+		
+		struct {
+			struct fi_msg_atomic msg;
+			struct fi_ioc msg_iov[SOCK_EP_MAX_IOV_LIMIT];
+			struct fi_rma_ioc rma_iov[SOCK_EP_MAX_IOV_LIMIT];
+			struct fi_ioc comparev[SOCK_EP_MAX_IOV_LIMIT];
+			size_t compare_count;
+			struct fi_ioc resultv[SOCK_EP_MAX_IOV_LIMIT];
+			size_t result_count;
+		} atomic;
+	} op;
+};
+
 struct sock_cntr {
 	struct fid_cntr cntr_fid;
 	struct sock_domain *domain;
@@ -206,6 +243,9 @@ struct sock_cntr {
 	struct dlist_entry rx_list;
 	struct dlist_entry tx_list;
 	fastlock_t list_lock;
+
+	fastlock_t trigger_lock;
+	struct dlist_entry trigger_list;
 
 	struct fid_wait *waitset;
 	int signal;
@@ -822,6 +862,7 @@ struct sock_conn_req_handle {
 	struct sock_conn_req *req;
 };
 
+
 int sock_verify_info(struct fi_info *hints);
 int sock_verify_fabric_attr(struct fi_fabric_attr *attr);
 int sock_verify_domain_attr(struct fi_domain_attr *attr);
@@ -1028,5 +1069,36 @@ ssize_t sock_comm_peek(struct sock_conn *conn, void *buf, size_t len);
 ssize_t sock_comm_discard(struct sock_conn *conn, size_t len);
 ssize_t sock_comm_data_avail(struct sock_conn *conn);
 ssize_t sock_comm_flush(struct sock_conn *conn);
+
+ssize_t sock_ep_recvmsg(struct fid_ep *ep, const struct fi_msg *msg, 
+			uint64_t flags);
+ssize_t sock_ep_sendmsg(struct fid_ep *ep, const struct fi_msg *msg, 
+			uint64_t flags);
+ssize_t sock_ep_trecvmsg(struct fid_ep *ep, 
+			 const struct fi_msg_tagged *msg, uint64_t flags);
+ssize_t sock_ep_tsendmsg(struct fid_ep *ep, 
+			 const struct fi_msg_tagged *msg, uint64_t flags);
+ssize_t sock_ep_rma_readmsg(struct fid_ep *ep, const struct fi_msg_rma *msg, 
+			    uint64_t flags);
+ssize_t sock_ep_rma_writemsg(struct fid_ep *ep, const struct fi_msg_rma *msg, 
+			     uint64_t flags);
+ssize_t sock_ep_tx_atomic(struct fid_ep *ep, 
+			  const struct fi_msg_atomic *msg, 
+			  const struct fi_ioc *comparev, void **compare_desc, 
+			  size_t compare_count, struct fi_ioc *resultv, 
+			  void **result_desc, size_t result_count, uint64_t flags);
+
+
+ssize_t sock_queue_rma_op(struct fid_ep *ep, const struct fi_msg_rma *msg, 
+			  uint64_t flags, uint8_t op_type);
+ssize_t sock_queue_atomic_op(struct fid_ep *ep, const struct fi_msg_atomic *msg, 
+			     const struct fi_ioc *comparev, size_t compare_count, 
+			     struct fi_ioc *resultv, size_t result_count, 
+			     uint64_t flags, uint8_t op_type);
+ssize_t sock_queue_tmsg_op(struct fid_ep *ep, const struct fi_msg_tagged *msg,
+			   uint64_t flags, uint8_t op_type);
+ssize_t sock_queue_msg_op(struct fid_ep *ep, const struct fi_msg *msg, 
+			  uint64_t flags, uint8_t op_type);
+
 
 #endif

--- a/prov/sockets/src/sock_atomic.c
+++ b/prov/sockets/src/sock_atomic.c
@@ -59,11 +59,11 @@
 #define SOCK_LOG_INFO(...) _SOCK_LOG_INFO(FI_LOG_EP_DATA, __VA_ARGS__)
 #define SOCK_LOG_ERROR(...) _SOCK_LOG_ERROR(FI_LOG_EP_DATA, __VA_ARGS__)
 
-static ssize_t sock_ep_tx_atomic(struct fid_ep *ep, 
-				  const struct fi_msg_atomic *msg, 
-				  const struct fi_ioc *comparev, void **compare_desc, 
-				  size_t compare_count, struct fi_ioc *resultv, 
-				  void **result_desc, size_t result_count, uint64_t flags)
+ssize_t sock_ep_tx_atomic(struct fid_ep *ep, 
+			  const struct fi_msg_atomic *msg, 
+			  const struct fi_ioc *comparev, void **compare_desc, 
+			  size_t compare_count, struct fi_ioc *resultv, 
+			  void **result_desc, size_t result_count, uint64_t flags)
 {
 	int i, ret;
 	size_t datatype_sz;
@@ -111,6 +111,13 @@ static ssize_t sock_ep_tx_atomic(struct fid_ep *ep,
 	if (sock_ep_is_write_cq_low(&tx_ctx->comp, flags)) {
 		SOCK_LOG_ERROR("CQ size low\n");
 		return -FI_EAGAIN;
+	}
+
+	if ((flags & FI_TRIGGER) &&
+	    (ret = sock_queue_atomic_op(ep, msg, comparev, compare_count,
+					resultv, result_count, flags, 
+					SOCK_OP_ATOMIC)) != 1) {
+		return ret;
 	}
 
 	src_len = 0;

--- a/prov/sockets/src/sock_cntr.c
+++ b/prov/sockets/src/sock_cntr.c
@@ -43,6 +43,9 @@
 #include "sock.h"
 #include "sock_util.h"
 
+#define SOCK_LOG_INFO(...) _SOCK_LOG_INFO(FI_LOG_EP_DATA, __VA_ARGS__)
+#define SOCK_LOG_ERROR(...) _SOCK_LOG_ERROR(FI_LOG_EP_DATA, __VA_ARGS__)
+
 const struct fi_cntr_attr sock_cntr_attr = {
 	.events = FI_CNTR_EVENTS_COMP,
 	.wait_obj = FI_WAIT_MUTEX_COND,
@@ -76,6 +79,80 @@ int sock_cntr_progress(struct sock_cntr *cntr)
 	return 0;
 }
 
+void sock_cntr_check_trigger_list(struct sock_cntr *cntr)
+{
+	struct sock_trigger *trigger;
+	struct dlist_entry *entry;
+	int ret = 0;
+
+	fastlock_acquire(&cntr->trigger_lock);
+	for (entry = cntr->trigger_list.next;
+	     entry != &cntr->trigger_list;) {
+		
+		trigger = container_of(entry, struct sock_trigger, entry);
+		entry = entry->next;
+
+		if (atomic_get(&cntr->value) < trigger->threshold)
+			continue;
+
+		switch (trigger->op_type) {
+		case SOCK_OP_SEND:
+			ret = sock_ep_sendmsg(trigger->ep, &trigger->op.msg.msg, 
+					trigger->flags);
+			break;
+
+		case SOCK_OP_RECV:
+			ret = sock_ep_recvmsg(trigger->ep, &trigger->op.msg.msg,
+					trigger->flags);
+			break;
+
+		case SOCK_OP_TSEND:
+			ret = sock_ep_tsendmsg(trigger->ep, &trigger->op.tmsg.msg, 
+					trigger->flags);
+			break;
+
+		case SOCK_OP_TRECV:
+			ret = sock_ep_trecvmsg(trigger->ep, &trigger->op.tmsg.msg, 
+					trigger->flags);
+			break;
+
+		case SOCK_OP_WRITE:
+			ret = sock_ep_rma_writemsg(trigger->ep, &trigger->op.rma.msg,
+					     trigger->flags);
+			break;
+
+		case SOCK_OP_READ:
+			ret = sock_ep_rma_readmsg(trigger->ep, &trigger->op.rma.msg,
+					     trigger->flags);
+			break;
+
+		case SOCK_OP_ATOMIC:
+			ret = sock_ep_tx_atomic(trigger->ep, &trigger->op.atomic.msg,
+					  trigger->op.atomic.comparev, 
+					  NULL,
+					  trigger->op.atomic.compare_count,
+					  trigger->op.atomic.resultv,
+					  NULL,
+					  trigger->op.atomic.result_count,
+					  trigger->flags);
+			break;
+			
+		default:
+			SOCK_LOG_ERROR("unsupported op\n");
+			ret = 0;
+			break;
+		}
+
+		if (ret != -FI_EAGAIN) {
+			dlist_remove(&trigger->entry);
+			free(trigger);
+		} else {
+			break;
+		}
+	}	
+	fastlock_release(&cntr->trigger_lock);
+}
+
 static uint64_t sock_cntr_read(struct fid_cntr *cntr)
 {
 	struct sock_cntr *_cntr;
@@ -91,6 +168,7 @@ int sock_cntr_inc(struct sock_cntr *cntr)
 	if (atomic_get(&cntr->value) >= atomic_get(&cntr->threshold))
 		pthread_cond_signal(&cntr->cond);
 	pthread_mutex_unlock(&cntr->mut);
+	sock_cntr_check_trigger_list(cntr);
 	return 0;
 }
 
@@ -113,6 +191,7 @@ static int sock_cntr_add(struct fid_cntr *cntr, uint64_t value)
 	if (atomic_get(&_cntr->value) >= atomic_get(&_cntr->threshold))
 		pthread_cond_signal(&_cntr->cond);
 	pthread_mutex_unlock(&_cntr->mut);
+	sock_cntr_check_trigger_list(_cntr);
 	return 0;
 }
 
@@ -126,6 +205,7 @@ static int sock_cntr_set(struct fid_cntr *cntr, uint64_t value)
 	if (atomic_get(&_cntr->value) >= atomic_get(&_cntr->threshold))
 		pthread_cond_signal(&_cntr->cond);
 	pthread_mutex_unlock(&_cntr->mut);
+	sock_cntr_check_trigger_list(_cntr);
 	return 0;
 }
 
@@ -160,7 +240,7 @@ static int sock_cntr_wait(struct fid_cntr *cntr, uint64_t threshold, int timeout
 		while (atomic_get(&_cntr->value) < threshold) {
 			sock_cntr_progress(_cntr);
 			if (timeout >= 0 && fi_gettime_ms() >= end_ms) {
-				ret = -FI_ETIMEDOUT;
+				ret = FI_ETIMEDOUT;
 				break;
 			}
 		}
@@ -172,6 +252,7 @@ static int sock_cntr_wait(struct fid_cntr *cntr, uint64_t threshold, int timeout
 	_cntr->is_waiting = 0;
 	atomic_set(&_cntr->threshold, ~0);
 	pthread_mutex_unlock(&_cntr->mut);
+	sock_cntr_check_trigger_list(_cntr);
 	return -ret;
 }		
 
@@ -232,6 +313,7 @@ static int sock_cntr_close(struct fid *fid)
 	
 	pthread_mutex_destroy(&cntr->mut);
 	fastlock_destroy(&cntr->list_lock);
+	fastlock_destroy(&cntr->trigger_lock);
 
 	pthread_cond_destroy(&cntr->cond);
 	atomic_dec(&cntr->domain->ref);
@@ -366,6 +448,9 @@ int sock_cntr_open(struct fid_domain *domain, struct fi_cntr_attr *attr,
 
 	dlist_init(&_cntr->tx_list);
 	dlist_init(&_cntr->rx_list);
+
+	dlist_init(&_cntr->trigger_list);
+	fastlock_init(&_cntr->trigger_lock);
 
 	_cntr->cntr_fid.fid.fclass = FI_CLASS_CNTR;
 	_cntr->cntr_fid.fid.context = context;

--- a/prov/sockets/src/sock_msg.c
+++ b/prov/sockets/src/sock_msg.c
@@ -59,10 +59,10 @@
 #define SOCK_LOG_INFO(...) _SOCK_LOG_INFO(FI_LOG_EP_DATA, __VA_ARGS__)
 #define SOCK_LOG_ERROR(...) _SOCK_LOG_ERROR(FI_LOG_EP_DATA, __VA_ARGS__)
 
-static ssize_t sock_ep_recvmsg(struct fid_ep *ep, const struct fi_msg *msg, 
-				uint64_t flags)
+ssize_t sock_ep_recvmsg(struct fid_ep *ep, const struct fi_msg *msg, 
+			uint64_t flags)
 {
-	int i;
+	int i, ret;
 	struct sock_rx_ctx *rx_ctx;
 	struct sock_rx_entry *rx_entry;
 	struct sock_ep *sock_ep;
@@ -93,6 +93,11 @@ static ssize_t sock_ep_recvmsg(struct fid_ep *ep, const struct fi_msg *msg,
 	if (sock_ep_is_recv_cq_low(&rx_ctx->comp, flags)) {
 		SOCK_LOG_ERROR("CQ size low\n");
 		return -FI_EAGAIN;
+	}
+
+	if ((flags & FI_TRIGGER) &&
+	    (ret = sock_queue_msg_op(ep, msg, flags, SOCK_OP_RECV)) != 1) {
+		return ret;
 	}
 
 	if (flags & FI_PEEK) {
@@ -165,8 +170,8 @@ static ssize_t sock_ep_recvv(struct fid_ep *ep, const struct iovec *iov,
 	return sock_ep_recvmsg(ep, &msg, SOCK_USE_OP_FLAGS);
 }
 
-static ssize_t sock_ep_sendmsg(struct fid_ep *ep, const struct fi_msg *msg, 
-				uint64_t flags)
+ssize_t sock_ep_sendmsg(struct fid_ep *ep, const struct fi_msg *msg, 
+			uint64_t flags)
 {
 	int ret, i;
 	uint64_t total_len;
@@ -217,6 +222,11 @@ static ssize_t sock_ep_sendmsg(struct fid_ep *ep, const struct fi_msg *msg,
 	if (sock_ep_is_send_cq_low(&tx_ctx->comp, flags)) {
 		SOCK_LOG_ERROR("CQ size low\n");
 		return -FI_EAGAIN;
+	}
+
+	if ((flags & FI_TRIGGER) &&
+	    (ret = sock_queue_msg_op(ep, msg, flags, SOCK_OP_SEND)) != 1) {
+		return ret;
 	}
 
 	memset(&tx_op, 0, sizeof(struct sock_op));
@@ -377,10 +387,10 @@ struct fi_ops_msg sock_ep_msg_ops = {
 	.injectdata = sock_ep_injectdata
 };
 
-static ssize_t sock_ep_trecvmsg(struct fid_ep *ep, 
-				 const struct fi_msg_tagged *msg, uint64_t flags)
+ssize_t sock_ep_trecvmsg(struct fid_ep *ep, 
+			 const struct fi_msg_tagged *msg, uint64_t flags)
 {
-	int i;
+	int i, ret;
 	struct sock_rx_ctx *rx_ctx;
 	struct sock_rx_entry *rx_entry;
 	struct sock_ep *sock_ep;
@@ -412,6 +422,11 @@ static ssize_t sock_ep_trecvmsg(struct fid_ep *ep,
 	if (sock_ep_is_recv_cq_low(&rx_ctx->comp, flags)) {
 		SOCK_LOG_ERROR("CQ size low\n");
 		return -FI_EAGAIN;
+	}
+
+	if ((flags & FI_TRIGGER) &&
+	    (ret = sock_queue_tmsg_op(ep, msg, flags, SOCK_OP_TRECV)) != 1) {
+		return ret;
 	}
 
 	if (flags & FI_PEEK) {
@@ -492,8 +507,8 @@ static ssize_t sock_ep_trecvv(struct fid_ep *ep, const struct iovec *iov,
 	return sock_ep_trecvmsg(ep, &msg, SOCK_USE_OP_FLAGS);
 }
 
-static ssize_t sock_ep_tsendmsg(struct fid_ep *ep, 
-				 const struct fi_msg_tagged *msg, uint64_t flags)
+ssize_t sock_ep_tsendmsg(struct fid_ep *ep, 
+			 const struct fi_msg_tagged *msg, uint64_t flags)
 {
 	int ret, i;
 	uint64_t total_len;
@@ -541,6 +556,11 @@ static ssize_t sock_ep_tsendmsg(struct fid_ep *ep,
 	if (sock_ep_is_send_cq_low(&tx_ctx->comp, flags)) {
 		SOCK_LOG_ERROR("CQ size low\n");
 		return -FI_EAGAIN;
+	}
+
+	if ((flags & FI_TRIGGER) &&
+	    (ret = sock_queue_tmsg_op(ep, msg, flags, SOCK_OP_TSEND)) != 1) {
+		return ret;
 	}
 
 	memset(&tx_op, 0, sizeof(tx_op));

--- a/prov/sockets/src/sock_rma.c
+++ b/prov/sockets/src/sock_rma.c
@@ -59,9 +59,8 @@
 #define SOCK_LOG_INFO(...) _SOCK_LOG_INFO(FI_LOG_EP_DATA, __VA_ARGS__)
 #define SOCK_LOG_ERROR(...) _SOCK_LOG_ERROR(FI_LOG_EP_DATA, __VA_ARGS__)
 
-static ssize_t sock_ep_rma_readmsg(struct fid_ep *ep, 
-					const struct fi_msg_rma *msg, 
-					uint64_t flags)
+ssize_t sock_ep_rma_readmsg(struct fid_ep *ep, const struct fi_msg_rma *msg, 
+			    uint64_t flags)
 {
 	int ret, i;
 	struct sock_op tx_op;
@@ -103,22 +102,26 @@ static ssize_t sock_ep_rma_readmsg(struct fid_ep *ep,
 	if (!conn)
 		return -FI_EAGAIN;
 
-	total_len = sizeof(struct sock_op_send) + 
-		(msg->iov_count * sizeof(union sock_iov)) +
-		(msg->rma_iov_count * sizeof(union sock_iov));
-
-	sock_tx_ctx_start(tx_ctx);
-	if (rbfdavail(&tx_ctx->rbfd) < total_len) {
-		ret = -FI_EAGAIN;
-		goto err;
-	}
-
 	SOCK_EP_SET_TX_OP_FLAGS(flags);
 	if (flags & SOCK_USE_OP_FLAGS)
 		flags |= tx_ctx->attr.op_flags;	
 
 	if (sock_ep_is_read_cq_low(&tx_ctx->comp, flags)) {
 		SOCK_LOG_ERROR("CQ size low\n");
+		return -FI_EAGAIN;
+	}
+
+	if ((flags & FI_TRIGGER) && 
+	    (ret = sock_queue_rma_op(ep, msg, flags, SOCK_OP_READ)) != 1) {
+		return ret;
+	}
+
+	total_len = sizeof(struct sock_op_send) + 
+		(msg->iov_count * sizeof(union sock_iov)) +
+		(msg->rma_iov_count * sizeof(union sock_iov));
+	
+	sock_tx_ctx_start(tx_ctx);
+	if (rbfdavail(&tx_ctx->rbfd) < total_len) {
 		ret = -FI_EAGAIN;
 		goto err;
 	}
@@ -218,9 +221,8 @@ static ssize_t sock_ep_rma_readv(struct fid_ep *ep, const struct iovec *iov,
 	return sock_ep_rma_readmsg(ep, &msg, SOCK_USE_OP_FLAGS);
 }
 
-static ssize_t sock_ep_rma_writemsg(struct fid_ep *ep, 
-				     const struct fi_msg_rma *msg, 
-				     uint64_t flags)
+ssize_t sock_ep_rma_writemsg(struct fid_ep *ep, const struct fi_msg_rma *msg, 
+			     uint64_t flags)
 {
 	int ret, i;
 	struct sock_op tx_op;
@@ -269,6 +271,11 @@ static ssize_t sock_ep_rma_writemsg(struct fid_ep *ep,
 	if (sock_ep_is_write_cq_low(&tx_ctx->comp, flags)) {
 		SOCK_LOG_ERROR("CQ size low\n");
 		return -FI_EAGAIN;
+	}
+
+	if ((flags & FI_TRIGGER) && 
+	    (ret = sock_queue_rma_op(ep, msg, flags, SOCK_OP_WRITE)) != 1) {
+		return ret;
 	}
 
 	memset(&tx_op, 0, sizeof(struct sock_op));

--- a/prov/sockets/src/sock_trigger.c
+++ b/prov/sockets/src/sock_trigger.c
@@ -1,0 +1,220 @@
+/*
+ * Copyright (c) 2014-2015 Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#  include <config.h>
+#endif /* HAVE_CONFIG_H */
+
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "sock.h"
+#include "sock_util.h"
+
+#define SOCK_LOG_INFO(...) _SOCK_LOG_INFO(FI_LOG_EP_DATA, __VA_ARGS__)
+#define SOCK_LOG_ERROR(...) _SOCK_LOG_ERROR(FI_LOG_EP_DATA, __VA_ARGS__)
+
+ssize_t sock_queue_rma_op(struct fid_ep *ep, const struct fi_msg_rma *msg,
+			  uint64_t flags, uint8_t op_type)
+{
+	struct sock_cntr *cntr;
+	struct sock_trigger *trigger;
+	struct fi_triggered_context *trigger_context;
+	struct fi_trigger_threshold *threshold;
+
+	trigger_context = (struct fi_triggered_context*)msg->context;
+	if ((trigger_context->event_type != FI_TRIGGER_THRESHOLD) ||
+	    (flags & FI_INJECT))
+		return -FI_EINVAL;
+
+	threshold = &trigger_context->trigger.threshold;
+	cntr = container_of(threshold->cntr, struct sock_cntr, cntr_fid);
+	if (atomic_get(&cntr->value) >= threshold->threshold)
+		return 1;
+
+	trigger = calloc(1, sizeof *trigger);
+	if (!trigger)
+		return -FI_ENOMEM;
+	
+	trigger->threshold = threshold->threshold;
+	memcpy(&trigger->op.rma.msg, msg, sizeof *msg);
+	trigger->op.rma.msg.msg_iov = &trigger->op.rma.msg_iov[0];
+	trigger->op.rma.msg.rma_iov = &trigger->op.rma.rma_iov[0];
+	
+	memcpy(&trigger->op.rma.msg_iov[0], &msg->msg_iov[0],
+	       msg->iov_count * sizeof (struct iovec));
+	memcpy(&trigger->op.rma.rma_iov[0], &msg->rma_iov[0],
+	       msg->rma_iov_count * sizeof (struct fi_rma_iov));
+
+	trigger->op_type = op_type;
+	trigger->ep = ep;
+	trigger->flags = flags;
+	
+	fastlock_acquire(&cntr->trigger_lock);
+	dlist_insert_tail(&trigger->entry, &cntr->trigger_list);
+	fastlock_release(&cntr->trigger_lock);
+	return 0;
+}
+
+ssize_t sock_queue_msg_op(struct fid_ep *ep, const struct fi_msg *msg, 
+			  uint64_t flags, uint8_t op_type)
+{
+	struct sock_cntr *cntr;
+	struct sock_trigger *trigger;
+	struct fi_triggered_context *trigger_context;
+	struct fi_trigger_threshold *threshold;
+
+	trigger_context = (struct fi_triggered_context*)msg->context;
+	if ((trigger_context->event_type != FI_TRIGGER_THRESHOLD) ||
+	    (flags & FI_INJECT))
+		return -FI_EINVAL;
+
+	threshold = &trigger_context->trigger.threshold;
+	cntr = container_of(threshold->cntr, struct sock_cntr, cntr_fid);
+	if (atomic_get(&cntr->value) >= threshold->threshold)
+		return 1;
+
+	trigger = calloc(1, sizeof *trigger);
+	if (!trigger)
+		return -FI_ENOMEM;
+	
+	trigger->threshold = threshold->threshold;
+
+	memcpy(&trigger->op.msg.msg, msg, sizeof *msg);
+	trigger->op.msg.msg.msg_iov = &trigger->op.msg.msg.msg_iov[0];
+	memcpy((void*)&trigger->op.msg.msg.msg_iov[0], &msg->msg_iov[0],
+	       msg->iov_count * sizeof (struct iovec));
+
+	trigger->op_type = op_type;
+	trigger->ep = ep;
+	trigger->flags = flags;
+	
+	fastlock_acquire(&cntr->trigger_lock);
+	dlist_insert_tail(&trigger->entry, &cntr->trigger_list);
+	fastlock_release(&cntr->trigger_lock);
+	return 0;
+}
+
+ssize_t sock_queue_tmsg_op(struct fid_ep *ep, const struct fi_msg_tagged *msg,
+			   uint64_t flags, uint8_t op_type)
+{
+	struct sock_cntr *cntr;
+	struct sock_trigger *trigger;
+	struct fi_triggered_context *trigger_context;
+	struct fi_trigger_threshold *threshold;
+
+	trigger_context = (struct fi_triggered_context*)msg->context;
+	if ((trigger_context->event_type != FI_TRIGGER_THRESHOLD) ||
+	    (flags & FI_INJECT))
+		return -FI_EINVAL;
+
+	threshold = &trigger_context->trigger.threshold;
+	cntr = container_of(threshold->cntr, struct sock_cntr, cntr_fid);
+	if (atomic_get(&cntr->value) >= threshold->threshold)
+		return 1;
+
+	trigger = calloc(1, sizeof *trigger);
+	if (!trigger)
+		return -FI_ENOMEM;
+	
+	trigger->threshold = threshold->threshold;
+	
+	memcpy(&trigger->op.tmsg.msg, msg, sizeof *msg);
+	trigger->op.tmsg.msg.msg_iov = &trigger->op.tmsg.msg.msg_iov[0];
+	memcpy((void*)&trigger->op.tmsg.msg.msg_iov[0], &msg->msg_iov[0],
+	       msg->iov_count * sizeof (struct iovec));
+
+	trigger->op_type = op_type;
+	trigger->ep = ep;
+	trigger->flags = flags;
+	
+	fastlock_acquire(&cntr->trigger_lock);
+	dlist_insert_tail(&trigger->entry, &cntr->trigger_list);
+	fastlock_release(&cntr->trigger_lock);
+	return 0;
+}
+
+ssize_t sock_queue_atomic_op(struct fid_ep *ep, const struct fi_msg_atomic *msg, 
+			     const struct fi_ioc *comparev, size_t compare_count, 
+			     struct fi_ioc *resultv, size_t result_count, 
+			     uint64_t flags, uint8_t op_type)
+{
+	struct sock_cntr *cntr;
+	struct sock_trigger *trigger;
+	struct fi_triggered_context *trigger_context;
+	struct fi_trigger_threshold *threshold;
+
+	trigger_context = (struct fi_triggered_context*)msg->context;
+	if ((trigger_context->event_type != FI_TRIGGER_THRESHOLD) ||
+	    (flags & FI_INJECT))
+		return -FI_EINVAL;
+
+	threshold = &trigger_context->trigger.threshold;
+	cntr = container_of(threshold->cntr, struct sock_cntr, cntr_fid);
+	if (atomic_get(&cntr->value) >= threshold->threshold)
+		return 1;
+
+	trigger = calloc(1, sizeof *trigger);
+	if (!trigger)
+		return -FI_ENOMEM;
+	
+	trigger->threshold = threshold->threshold;
+	memcpy(&trigger->op.atomic.msg, msg, sizeof *msg);
+	trigger->op.atomic.msg.msg_iov = &trigger->op.atomic.msg_iov[0];
+	trigger->op.atomic.msg.rma_iov = &trigger->op.atomic.rma_iov[0];
+
+	memcpy(&trigger->op.atomic.msg_iov[0], &msg->msg_iov[0],
+	       msg->iov_count * sizeof (struct fi_ioc));
+	memcpy(&trigger->op.atomic.rma_iov[0], &msg->rma_iov[0],
+	       msg->iov_count * sizeof (struct fi_rma_ioc));
+
+	if (comparev) {
+		memcpy(&trigger->op.atomic.comparev[0], &comparev[0],
+		       compare_count * sizeof (struct fi_ioc));
+	}
+
+	if (resultv) {
+		memcpy(&trigger->op.atomic.resultv[0], &resultv[0],
+		       result_count * sizeof (struct fi_ioc));
+	}	
+	
+	trigger->op_type = op_type;
+	trigger->ep = ep;
+	trigger->flags = flags;
+	
+	fastlock_acquire(&cntr->trigger_lock);
+	dlist_insert_tail(&trigger->entry, &cntr->trigger_list);
+	fastlock_release(&cntr->trigger_lock);
+	return 0;
+}

